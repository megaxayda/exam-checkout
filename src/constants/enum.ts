export enum PizzaSize {
  SMALL = 'SMALL',
  MEDIUM = 'MEDIUM',
  LARGE = 'LARGE',
}

// customer:

// pizza type:

// price

// reduction type: free x | direct reduction | percent reduction | lowered price

// reduction num: x | price | %

// min num:

// priority:

export enum ReductionType {
  FREE_PER_X = 'FREE_PER_X',
  DIRECT = 'DIRECT',
  PERCENT = 'PERCENT',
  MODIFIED_PRICE = 'MODIFIED_PRICE',
}

export const reductionTypeOption = [
  {
    name: 'Free per X',
    value: ReductionType.FREE_PER_X,
  },
  {
    name: 'Direct',
    value: ReductionType.DIRECT,
  },
  {
    name: 'Percent',
    value: ReductionType.PERCENT,
  },
  {
    name: 'Modified Price',
    value: ReductionType.MODIFIED_PRICE,
  },
];
