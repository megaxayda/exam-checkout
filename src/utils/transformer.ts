import { Transformer } from 'dinero.js';

const transformer: Transformer<number> = (props) =>
  `${props.currency.code} ${props.amount}`;

export default transformer;
