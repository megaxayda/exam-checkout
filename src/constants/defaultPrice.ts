import { USD } from '@dinero.js/currencies';
import { dinero } from 'dinero.js';

import { PizzaSize } from './enum';

const defaultPrice = {
  [PizzaSize.SMALL]: dinero({ amount: 1199, currency: USD, scale: 2 }),
  [PizzaSize.MEDIUM]: dinero({ amount: 1599, currency: USD, scale: 2 }),
  [PizzaSize.LARGE]: dinero({ amount: 2199, currency: USD, scale: 2 }),
};

export default defaultPrice;
