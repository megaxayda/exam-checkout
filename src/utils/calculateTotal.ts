import { USD } from '@dinero.js/currencies';
import defaultPrice from 'constants/defaultPrice';
import { PizzaSize, ReductionType } from 'constants/enum';
import { allocate, dinero, multiply, subtract } from 'dinero.js';
import isEmpty from 'lodash/isEmpty';
import { PriceConfig } from 'slices/priceConfigSlice';

export const dineroFromFloat = (amount: number) => {
  const factor = USD.base ** USD.exponent;
  const a = Math.round(amount * factor);

  return dinero({ amount: a, currency: USD });
};

export const filterPriceConfigs = (
  priceConfigs: PriceConfig[],
  quantity: number,
): PriceConfig[] => {
  // FILTER BY MINIMUM REQUIRED
  const priceConfigsByMinimum = priceConfigs.filter(
    (config) => quantity >= config.minRequired,
  );

  // FILTER BY PRIORITY
  const maxPriority = Math.min(...priceConfigsByMinimum.map((config) => config.priority));
  const priceConfigsByPriority = priceConfigsByMinimum.filter(
    (config) => config.priority === maxPriority,
  );

  return priceConfigsByPriority;
};

const calculateTotal = (
  priceConfigs: PriceConfig[],
  size: PizzaSize,
  quantity: number,
) => {
  try {
    if (isNaN(quantity) || quantity === 0) {
      return { total: dineroFromFloat(0) };
    }

    const filteredConfigs = filterPriceConfigs(priceConfigs, quantity);

    let modifiedQuantity = quantity;
    const price = defaultPrice[size];
    let modifiedPrice = price;

    if (isEmpty(filteredConfigs)) {
      return { total: multiply(modifiedPrice, modifiedQuantity) };
    }

    for (const config of filteredConfigs) {
      const { reductionModifier, minRequired } = config;

      switch (config.reductionType) {
        case ReductionType.FREE_PER_X:
          if (minRequired === 0) {
            throw 'reductionModifier can not be 0';
          }
          modifiedQuantity = quantity - Math.floor(quantity / minRequired);
          break;

        // Only apply to original price
        case ReductionType.DIRECT:
          modifiedPrice = subtract(price, dineroFromFloat(reductionModifier));
          break;

        // Only apply to original price
        case ReductionType.PERCENT: {
          if (reductionModifier > 100) {
            throw 'reductionModifier can not be bigger than 100';
          }
          const [d1] = allocate(price, [
            100 - Math.floor(reductionModifier),
            Math.floor(reductionModifier),
          ]);
          modifiedPrice = d1;
          break;
        }

        case ReductionType.MODIFIED_PRICE: {
          modifiedPrice = dineroFromFloat(reductionModifier);
          break;
        }

        default:
          break;
      }
    }

    return { total: multiply(modifiedPrice, modifiedQuantity), configs: filteredConfigs };
  } catch (error) {
    console.error(error);
    alert && alert(String(error));
  }

  return { total: dineroFromFloat(0) };
};

export default calculateTotal;
