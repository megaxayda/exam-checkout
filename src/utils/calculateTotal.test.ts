import { PizzaSize, ReductionType } from 'constants/enum';
import { toFormat } from 'dinero.js';

import calculateTotal, { dineroFromFloat, filterPriceConfigs } from './calculateTotal';
import transformer from './transformer';

describe('calculateTotal', () => {
  const priceConfigs = [
    {
      name: 'amazon 50% if more than 10',
      customerType: 'amazon',
      pizzaSize: PizzaSize.SMALL,
      reductionType: ReductionType.PERCENT,
      reductionModifier: 50,
      minRequired: 10,
      priority: 1,
    },
    {
      name: 'amazon 20% for 1',
      customerType: 'amazon',
      pizzaSize: PizzaSize.SMALL,
      reductionType: ReductionType.PERCENT,
      reductionModifier: 20,
      minRequired: 1,
      priority: 2,
    },
  ];

  describe('dineroFromFloat', () => {
    test.each`
      input     | expected
      ${0}      | ${'USD 0'}
      ${2}      | ${'USD 2'}
      ${1.9}    | ${'USD 1.9'}
      ${2.99}   | ${'USD 2.99'}
      ${3.111}  | ${'USD 3.11'}
      ${4.6666} | ${'USD 4.67'}
    `('dineroFromFloat($input)', ({ input, expected }) => {
      expect(toFormat(dineroFromFloat(input), transformer)).toBe(expected);
    });
  });

  describe('filterPriceConfigs', () => {
    it('should return correct configs', () => {
      expect(filterPriceConfigs(priceConfigs, 1)).toMatchSnapshot();
    });
  });

  describe('calculateTotal', () => {
    it('should return USD 0 if quantity = 0', () => {
      const { total } = calculateTotal([], PizzaSize.SMALL, 0);

      expect(toFormat(total, transformer)).toBe('USD 0');
    });

    it('should return USD 0 if quantity is NaN', () => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      const { total } = calculateTotal([], PizzaSize.SMALL, 'abc');

      expect(toFormat(total, transformer)).toBe('USD 0');
    });

    it('should return default calculation if price configs is empty', () => {
      const { total } = calculateTotal([], PizzaSize.SMALL, 10);

      expect(toFormat(total, transformer)).toBe('USD 119.9');
    });

    it('should return correct price', () => {
      const { total, configs } = calculateTotal(priceConfigs, PizzaSize.SMALL, 10);

      expect(toFormat(total, transformer)).toBe('USD 60');
      expect(configs).toMatchSnapshot();
    });
  });
});
