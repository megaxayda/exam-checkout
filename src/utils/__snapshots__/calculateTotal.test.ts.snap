// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateTotal calculateTotal should return correct price 1`] = `
Array [
  Object {
    "customerType": "amazon",
    "minRequired": 10,
    "name": "amazon 50% if more than 10",
    "pizzaSize": "SMALL",
    "priority": 1,
    "reductionModifier": 50,
    "reductionType": "PERCENT",
  },
]
`;

exports[`calculateTotal filterPriceConfigs should return correct configs 1`] = `
Array [
  Object {
    "customerType": "amazon",
    "minRequired": 1,
    "name": "amazon 20% for 1",
    "pizzaSize": "SMALL",
    "priority": 2,
    "reductionModifier": 20,
    "reductionType": "PERCENT",
  },
]
`;
